railties/lib/rails/app_loader.rb

```diff
@@ -10,13 +10,19 @@ module AppLoader # :nodoc:
     BUNDLER_WARNING = <<EOS
 Looks like your app's ./bin/rails is a stub that was generated by Bundler.
 
-In Rails #{Rails::VERSION::MAJOR}, your app's bin/ directory contains executables that are versioned
+Beginning In Rails 4, your app's bin/ directory contains executables that are versioned
 like any other source code, rather than stubs that are generated on demand.
 
-Here's how to upgrade:
+If your application was created in Rails 4+, you might be seeing this message
+if your bundle install command is using the --binstubs option and thereby overriding the
+standard Rails bin/rails binstub. You might need to stop using the --binstubs option,
+or at least ensure that it is using a path other than ./bin. You may need to coordinate
+this change with your environment's PATH.
```

Comment:
Sure.  I was inadvertently running `bundle install --binstubs` on a production server as part of a deploy step. This would generate new binstubs and override the Rails binstub that was already in `./bin`, thus causing the bundle binstub error message to be shown even though my app already has the Rails binstub in `./bin` in the repo. For me the solution, was just _not_ to use `--binstubs`.

However, we probably don't want to suggest not using `--binstubs` as the only solution because there might be legitimate reasons to use that option. As far as I know, the Heroku buildpack still uses it, but with a non-default folder: `bundle install --without development:test --path vendor/bundle --binstubs vendor/bundle/bin -j4 --deployment`).  If the binstubs are placed somewhere besides `./bin`, Rails will see the correct Rails-generated `./bin/rails` binstub and not raise an error.  I'm presuming some people may need to modify their PATH if they place bundler binstubs in an alternate location (like `vendor/bundle/bin`), but perhaps I'm missing something.